### ISSUE TITLE
A concern about the loss function.

### ISSUE THREAD

**Issue by XuHQ1997** [czq142857/IM-NET-pytorch#13]
Thanks for the codes! 

But the loss function used here is confusing to me.
I notice that a (leaky) clamp operation is applied to the output of generator.
`l7 = torch.max(torch.min(l7, l7*0.01+0.99), l7*0.01)`
Then, MSE Loss is used. 

Here is the problem. Let's say a positive sample whose target is 1, then a prediction for it is in range of [0, 1), and the other one is negtative. Naturally, we expect that the loss of the former should be less than the latter. But via the loss function used here, that is not sure. There will be a local minimum in the negative area. Do you think this is gonna make it difficult to train?

By the way, is there any reason why you don't use BCE or NLL loss here? Or, any advise about the loss function?

Looking forward for your reply.

**Comment by czq142857** [comment]
![image](https://user-images.githubusercontent.com/24209790/120911172-ca09af80-c639-11eb-9cbc-d401c60fcca2.png)
The leaky clamp operation looks like this. As you see, the problem you described does not exist.

Personally, I prefer piece-wise linear activation functions, which is why I usually use leaky ReLU or leaky clamp. Since I use leaky clamp in the output layer (instead of sigmoid), there will be negative values in the outputs, therefore BCE cannot be applied.

**Comment by XuHQ1997** [comment]
Oh, sorry~ I just confused loss and gradient. 
The gradient may be greater when the prediction is closer to 1. Emm...It seems not a big problem, right?
![图片1](https://user-images.githubusercontent.com/74892500/120911944-bc960900-c6bd-11eb-944c-d86427da21fa.png)

Thanks for your reply.

**Comment by czq142857** [comment]
When computing the gradient, you need to consider the ground truth. The gradients you showed on the figure seem to assume the ground truth is 1.

It is actually better to consider the gradient as the product of the gradient from MSE and the constant gradient (1 or 0.01) from this piece-wise linear function.

**Comment by XuHQ1997** [comment]
Yes, you'r right. Because negative samples are much more than the positive samples, I guess the network would struggle to learn the positive samples. So I discussed about the positive samples. And I wonder whether the imbalance between positive and negative samples is the main reason for which we need train the network so long.

**Comment by czq142857** [comment]
I would consider the network structure as the main reason, i.e., the continuity/smoothness of the functions represented by MLPs.

**Comment by XuHQ1997** [comment]
Ok. Thanks for your patient :)
